collect_app/src/test/java/org/odk/collect/android/formentry/backgroundlocation/BackgroundLocationManagerTest.java

```diff
@@ -0,0 +1,361 @@
+package org.odk.collect.android.formentry.backgroundlocation;
+
+import android.location.Location;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+import org.odk.collect.android.location.LocationTestUtils;
+import org.odk.collect.android.location.client.FakeLocationClient;
+import org.odk.collect.android.logic.AuditConfig;
+import org.odk.collect.android.logic.AuditEvent;
+import org.robolectric.RobolectricTestRunner;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+import static org.mockito.Mockito.atLeastOnce;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(RobolectricTestRunner.class)
+public class BackgroundLocationManagerTest {
+    @Rule
+    public MockitoRule rule = MockitoJUnit.rule();
+
+    private BackgroundLocationManager backgroundLocationManager;
+
+    private FakeLocationClient fakeLocationClient;
+    @Mock
+    private BackgroundLocationHelper locationHelper;
+
+    @Before
+    public void setUp() {
+        fakeLocationClient = new FakeLocationClient();
+        backgroundLocationManager = new BackgroundLocationManager(fakeLocationClient, locationHelper);
+    }
+
+    /**
+     * activityDisplayed is called from both onStart and loadingComplete in FormEntryActivity. This
+     * test confirms that it doesn't matter how many times activityDisplayed is called until the
+     * form has been loaded.
+     */
+    @Test
+    public void displayingActivityManyTimes_ShouldHaveNoEffect_WhenFormNotLoaded() {
+        when(locationHelper.currentFormCollectsBackgroundLocation()).thenReturn(true);
+
+        when(locationHelper.arePlayServicesAvailable()).thenReturn(true);
+        when(locationHelper.isBackgroundLocationPreferenceEnabled()).thenReturn(true);
+
+        assertThat(backgroundLocationManager.activityDisplayed(), is(nullValue()));
+        assertThat(backgroundLocationManager.isPendingPermissionCheck(), is(false));
+
+        assertThat(backgroundLocationManager.activityDisplayed(), is(nullValue()));
+        assertThat(backgroundLocationManager.isPendingPermissionCheck(), is(false));
+
+        verify(locationHelper, never()).logAuditEvent(Mockito.any());
+
+        backgroundLocationManager.formFinishedLoading();
+        backgroundLocationManager.activityDisplayed();
+
+        assertThat(backgroundLocationManager.isPendingPermissionCheck(), is(true));
+        verify(locationHelper, atLeastOnce()).logAuditEvent(Mockito.any());
+    }
+
+    @Test
+    public void locationPermissions_ShouldBeRequested_WhenFormRequestsBackgroundLocation_AndPreconditionsAreMet() {
```

Comment:
I found the test names here a little hard to read. Always hard to get to a readable state with JUnit methods but maybe they could be more focused with the method under test (like an `<subject>_<condition>_<expectation>` style) like:

`formFinishedLoading_whenPreconditionsAreMet_makesIsPendingPermissionCheckReturnTrue`

Maybe that's not better?

Reply:
🤷🏻‍♀️I think I was consistent with `<subject>_<expectation>_<most important conditions>`. I tried to make `subject` more about the thing that the user will experience.